File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.java
import java.util.*;
class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
int answer = 0;
      HashMap<String,Integer> w = new HashMap<>();
      
      for(int j=0; j<want.length; j++){
        w.put(want[j], number[j]);
      }

      for(int i=0; i<discount.length-9; i++){
        if(dayCount(i, discount, w)){
          answer++;
        }
      }
      
        return answer;
    }
    
    public boolean dayCount(int start, String[] discount, HashMap<String, Integer> w){
      int left = start;
      int right = start+9;
      HashMap<String,Integer> ch = new HashMap<>();
      for(int i=left; i<=right; i++){
        ch.put(discount[i], ch.getOrDefault(discount[i], 0)+1);
      }

      for(String s : w.keySet()){
        if(!ch.containsKey(s) || w.get(s) != ch.get(s)){
          return false;
        }
      }

      return true;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
// 10일 연속 일치할 경우에 맞춰서 회원가입 진행
function solution(want, number, discount) {
    let answer = 0;
    
    const map = new Map();
    for (let i = 0; i < want.length; i++) {
        map.set(want[i], number[i]);
    }
    
    for (let i = 0; i <= discount.length - 10; i++) {
        const copy = new Map([...map]);
        
        let count = 0;
        for (let j = i; j < i + 10; j++) {
            const target = discount[j];
            
            if (!copy.has(target)) break;
            if (!copy.get(target)) break;
            
            copy.set(target, copy.get(target) - 1);
            count++;
        }
        
        if (count === 10) answer++;
    }
    return answer;
}